protocols/gossipsub/src/behaviour.rs

```diff
@@ -1460,85 +1517,137 @@ where
-                let mut msg = GossipsubMessageWithId::new(msg, msg_id.clone());
-                self.mesage_with_id_is_valid(&mut msg, propagation_source);
+                let msg_id = msg_id.clone();
+                self.message_is_valid(&msg_id, &mut raw_message, propagation_source);
                 if let Some((peer_score, ..)) = &mut self.peer_score {
-                    peer_score.duplicated_message(propagation_source, &msg);
+                    peer_score.duplicated_message(propagation_source, &msg_id, &raw_message.topic);
                 }
                 return;
             }
         }
-        let msg = GenericGossipsubMessage::from(msg);
-        let msg_id = self.config.message_id(&msg);
-        let mut msg = GossipsubMessageWithId::new(msg, msg_id);
 
-        if !self.mesage_with_id_is_valid(&mut msg, propagation_source) {
-            return;
-        }
+        // Try and decompress the message. If it fails, consider it invalid.
+        let message = match GossipsubMessage::from_raw(
+            self.message_compression,
+            raw_message.clone(),
+            self.config.max_transmit_size(),
+        ) {
+            Ok(message) => message,
+            Err(e) => {
+                debug!("Invalid message. Decompression error: {:?}", e);
+                // Reject the message and return
+                self.handle_invalid_message(
+                    propagation_source,
+                    raw_message,
+                    ValidationError::DecompressionFailed,
+                );
+                return;
+            }
+        };
 
-        // Add the message to the duplicate caches and memcache.
+        let msg_id = self.config.message_id(&message);
+
+        // Add the message to the duplicate caches
         if let Some(fast_message_id) = fast_message_id {
             // add id to cache
             self.fast_messsage_id_cache
                 .entry(fast_message_id)
-                .or_insert_with(|| msg.message_id().clone());
+                .or_insert_with(|| msg_id.clone());
         }
-        if !self.duplicate_cache.insert(msg.message_id().clone()) {
+        if !self.duplicate_cache.insert(msg_id.clone()) {
             debug!(
                 "Message already received, ignoring. Message: {}",
-                msg.message_id()
+                msg_id.clone()
             );
             if let Some((peer_score, ..)) = &mut self.peer_score {
-                peer_score.duplicated_message(propagation_source, &msg);
+                peer_score.duplicated_message(propagation_source, &msg_id, &message.topic);
             }
             return;
         }
         debug!(
             "Put message {:?} in duplicate_cache and resolve promises",
-            msg.message_id()
+            msg_id
         );
 
+        // Check the validity of the message
+        if !self.message_is_valid(&msg_id, &mut raw_message, propagation_source) {
+            return;
+        }
```

Comment:
Any particular reason why this got moved after the duplicate cache check? Do you think this was a bug before or does it have to do with the changes?

Reply:
Mainly wanted to prevent all future invalids from being processed. Simply drop them as duplicates.